### DISCUSSION TITLE
[EventScripter] prop:categories_comma_separated parameters? or splice, reduce value ?

### DISCUSSION THREAD

**Discussion by jonlepage** [mgpai/resources#1] [Q&A] [answered]
hi @mgpai and thanks for your time.
The script seem not work on jdowload event
```java
File name -> contains -> (.+\d+p\.\[[^,]+,[^,]+,[^,]+,[^,]+),.+(\s\]\..+)
Downloadurl - > contains -> pornhub.com
Then set -> filename -> <jd:orgfilename:1><jd:orgfilename:2>
```
<details>
  <summary>i tested on this link (**spoil**)</summary>
https://www.pornhub.com/model/brunettviphotass/videos
</details>

It give me this !?
![image](https://user-images.githubusercontent.com/24865815/109844462-25329d00-7c1a-11eb-9751-2e036b73f83b.png)

this is my actual approach and format i would like get.
```java
File name -> contains -> (.+?(?=_))
Downloadurl - > contains -> pornhub.com
Then set -> packageName -> <jd:prop:username>
Then set -> filename -> <jd:prop:date>.<jd:orgfilename:1>.<jd:prop:username>.<jd:prop:quality>p.[ <jd:prop:categories_comma_separated> ].<jd:orgfiletype>
```
![image](https://user-images.githubusercontent.com/24865815/109845458-2d3f0c80-7c1b-11eb-9beb-8639ff79101d.png)

The only issue there, is `<jd:prop:categories_comma_separated>` will fetching all tags and it can be very long.
My Os is limited to 260 characters.

Am thinking about 2 solution.
The first , i think it will the best, if you have a plugin or script to do this ?
Find a way to add `metaData` in video file, similar like when you rip a Blu-ray collection, where you can fill `meta-tag` for your OS search engine.
I made a test and when i edit this manually, my os (window) actually take those keyword in the search engine.
![image](https://user-images.githubusercontent.com/24865815/109845986-b2c2bc80-7c1b-11eb-8082-e94c58a6b4ec.png)

But am not sure how preform this in Jdownloader environment because the js engine are very limited ! 
Can we call `nodejs` on local machine with eventScripter to pass variables and make task for add meta in the file ?
I also try the `comment` section but seem do nothing ! i suppose it only for jdownloader internal?
![image](https://user-images.githubusercontent.com/24865815/109847911-b48d7f80-7c1d-11eb-965b-8e77fbf6fed1.png)
 
 The other solution will be try a splice value from  `<jd:prop:categories_comma_separated>` to limited on 4 first tag, but i dont know how perform this.
 thanks for you time and help

**Comment by mgpai** [comment]
Tested with your link. Working fine.

![image](https://user-images.githubusercontent.com/23146220/109850448-6f337700-7c4a-11eb-887c-58448fdb0cbe.png)

**Comment by mgpai** [comment, marked as answer]
Have you removed/disabled your rule by any chance? This rule parses the file name generated by your rule. Your rule should be active and my rule should be placed after that.

**Reply by jonlepage** [reply]
interesting, good to know ! 
but nop i dont disabled my script.
it weird working on your side and not me .
am trying understand ! sry am noob on jsdownload.
![image](https://user-images.githubusercontent.com/24865815/109853058-7a26e100-7c23-11eb-959b-3c9aea0c5636.png)

**Reply by jonlepage** [reply]
ho wait i see you dont use the preview button ! 
![image](https://user-images.githubusercontent.com/24865815/109853386-d68a0080-7c23-11eb-80ff-09162777dcf8.png)

**Reply by jonlepage** [reply]
okey is work fine , this is awesome thanks a lot for your help.
I have another question ?

Do jdownloader have a `d.ts` where i can add inside a vscode project for get Intelisence and documentation ? 
how do you code your script ? you code blind?😮

**Reply by jonlepage** [reply]
i look in this list and i cant found nothing about the js api of jdownloader, 
https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types

**Reply by mgpai** [reply]
> Do jdownloader have a `d.ts` where i can add inside a vscode project for get Intelisence and documentation ?


No. 

> how do you code your script ? you code blind?😮


You can use 'alert' method to output to dialog box, or 'log' to log it to stderr and to JDownloader log Files, or use 'writeFile' to output to a text file with specified path.

**Reply by mgpai** [reply]
> i look in this list and i cant found nothing about the js api of jdownloader,
> https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types

The only reference you will find is in the 'show/hide help' option in the main men of the eventscripter script editor. It contains a list  of all the objects methods/properties which can be used in the script.

**Reply by jonlepage** [reply]
> > Do jdownloader have a `d.ts` where i can add inside a vscode project for get Intelisence and documentation ?
> 
> No.
> 
> > how do you code your script ? you code blind?😮
> 
> You can use 'alert' method to output to dialog box, or 'log' to log it to stderr and to JDownloader log Files, or use 'writeFile' to output to a text file with specified path.

ho oki thanks i will try, it's even more hellish than debugging with console.log everywhere. 😂

btw i dont know java, but it seem possible to convert java => ts  => d.ts for js.
It would be great to see this kind of support one day :)
https://stackoverflow.com/questions/32451966/generate-typescript-interfaces-from-java-interfaces

I will not bother you longer, thank you a lot for your help and your time it was very kind.
@+

**Reply by mgpai** [reply]
You're Welcome. :)
